crates/core/src/live_socket/mod.rs

```diff
@@ -97,32 +102,79 @@ impl LiveChannel {
 
 #[cfg_attr(not(target_family = "wasm"), uniffi::export(async_runtime = "tokio"))]
 impl LiveChannel {
+    pub fn document(&self) -> FFiDocument {
+        self.document.clone()
+    }
     pub fn channel(&self) -> Arc<Channel> {
         self.channel.clone()
     }
+    pub fn set_event_handler(&self, handler: Box<dyn DocumentChangeHandler>) {
+        self.document.set_event_handler(handler);
+    }
 
     pub async fn merge_diffs(&self) -> Result<(), LiveSocketError> {
         // TODO: This should probably take the event closure to send changes back to swift/kotlin
-        let mut document = self.document.clone();
+        let document = self.document.clone();
         let events = self.channel.events();
         loop {
             let event = events.event().await?;
-            if let Event::User { user: user_event } = event.event {
-                if user_event == "diff" {
-                    let payload = event.payload.to_string();
-                    debug!("PAYLOAD: {payload}");
-                    let diff: RootDiff = serde_json::from_str(payload.as_str())?;
-                    debug!("diff: {diff:#?}");
-                    document = document.merge(diff)?;
+            match event.event {
+                Event::Phoenix { .. } => {
+                    todo!();
                 }
-            }
+                Event::User { user } => {
+                    if user == "diff" {
+                        let payload = event.payload.to_string();
+                        debug!("PAYLOAD: {payload}");
+                        // This function merges and uses the event handler set in `set_event_handler`
+                        // which will call back into the Swift/Kotlin.
+                        document.merge_fragment_json(payload)?;
+                    }
+                }
+            };
```

Comment:
I think we'll want this in the future but it's definitely not fleshed out now. A reasonable stepping stone is to use the channel and pass those back into the Document via `merge_fragment_json` on the swift side for now.